Permit re-setting --object-bits

tautschnig opened the pull request:
Support choosing the number of object bits to be used by CBMC at runtime even with pre-compiled goto binaries.

Fixes: #5443

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Permit re-setting --object-bits

Support choosing the number of object bits to be used by CBMC at runtime
even with pre-compiled goto binaries. Update the goto-cc section in the
CPROVER manual to remove any documentation of object-bits (it remains
documented in the "Memory and pointers in CBMC" section).

Fixes: #5443

tautschnig committed:
Nondet-static: also update symbol table

This is to make sure that that re-creating initialisation functions
restores non-deterministic values.

kroening commented:
The title of the PR is a bit misleading -- this _moves_ the setting of object bits from goto-cc (or the front-end) to the solver.

This is the right thing to do (the only better thing is to get rid of object bits altogether).

This is a change of user-visible behaviour, and needs to be advertised appropriately.

tautschnig commented:
> The title of the PR is a bit misleading -- this _moves_ the setting of object bits from goto-cc (or the front-end) to the solver.
> 

I'm all up for changing the title -- I just thought a key problem that, with the previous implementation, GOTO binaries had their object bits firmly set. Now, one can use different values with `cbmc --object-bits ...` on a single GOTO binary. 

> This is the right thing to do (the only better thing is to get rid of object bits altogether).
> 
> This is a change of user-visible behaviour, and needs to be advertised appropriately.

Yes, this is user-visible -- but we hadn't really documented that undesirable need of setting `--object-bits` via `goto-cc`it seems?! Any suggestions on how to best document this change?

kroening commented:
> > The title of the PR is a bit misleading -- this _moves_ the setting of object bits from goto-cc (or the front-end) to the solver.
> 
> I'm all up for changing the title -- I just thought a key problem that, with the previous implementation, GOTO binaries had their object bits firmly set. Now, one can use different values with `cbmc --object-bits ...` on a single GOTO binary.

After the change, is it still possible to set object-bits via goto-cc? If not, then it's moving!

> Yes, this is user-visible -- but we hadn't really documented that undesirable need of setting `--object-bits` via `goto-cc`it seems?! Any suggestions on how to best document this change?

It may be helpful for users to keep a note in the man page for a while, saying that the option is now given to `cdmc`. Given that the option was always considered surprising, I'd be ok to make the change without a major version increase.

martin-cs reviewed (approved):
The small amount that touched things I am "owner" for seems to be correctly implemented. As I understand it this should be a back-end option so yes, setting it in `cbmc` not `goto-cc` makes sense.

tautschnig commented:
> > Yes, this is user-visible -- but we hadn't really documented that undesirable need of setting `--object-bits` via `goto-cc`it seems?! Any suggestions on how to best document this change?
> 
> It may be helpful for users to keep a note in the man page for a while, saying that the option is now given to `cdmc`. Given that the option was always considered surprising, I'd be ok to make the change without a major version increase.

I have now added a section "BACKWARD COMPATIBILITY" to the man page.

kroening commented on src/ansi-c/library/cprover.h:
```diff
@@ -26,8 +26,8 @@ extern const void *__CPROVER_deallocated;
 extern int __CPROVER_malloc_failure_mode;
-extern __CPROVER_size_t __CPROVER_max_malloc_size;
 extern __CPROVER_bool __CPROVER_malloc_may_fail;
+extern __CPROVER_thread_local __CPROVER_size_t __CPROVER_max_malloc_size;
```
Why is this thread local?

tautschnig replied:
This is to avoid our concurrency encoding picking the variable as one that can't be constant-propagated.

tautschnig replied:
I have now added a comment to clarify this.

kroening replied:
Would making it `const` achieve that?

tautschnig replied:
We don't commonly trust `const` at this time. We could, however, consider a lightweight analysis that establishes that a variable (irrespective of whether it is marked `const` or not) is only ever written once (and doesn't have its address taken).

thomasspriggs commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -99,11 +100,11 @@ static void nondet_static(
+          sym.type(), instruction.source_location()};
+        instruction.assign_rhs_nonconst() = nondet;
+        goto_model.symbol_table.get_writeable_ref(sym.get_identifier()).value =
+          nondet;
```
This line here seems to be the cause of the failure in the incremental SMT2 decision procedure. The failure happens during the conversion of the `y` symbol. Before your PR, the symbol value is an expression for the integer 23. After your PR it gains the value 
```
side_effect
  * type: signedbv
      * width: 32
      * #c_type: signed_int
  * #source_location: 
    * file: main.c
    * line: 4
    * function: 
    * working_directory: /home/tspriggs/Development/cbmc/regression/cbmc/no_nondet_static
  * statement: nondet
  * is_nondet_nullable: 1
```

This isn't supported by the decision procedure. I think the expressions sent to it either directly or from looking up the values of symbols are expected to be side effect free. The only exception would be in the case where we take the address of a code typed expression. The example in this test does not fall into this exception, because `y` symbol_exprt has a bit vector type.

It could be argued that the decision procedure should not error in this case. However I would say that this change in the value of symbols unrelated to max_malloc_size / object bits appears to be a change unrelated to the intended purpose of this PR.

tautschnig replied:
I need to figure out why the value in the symbol table is used directly. What should happen is that such a value is compiled into an assignment in `__CPROVER_initialize`, and `replace_nondet` will resolve this during symex.

tautschnig replied:
The problem appears to have been introduced in f286c07c07d: @thomasspriggs, why is it deemed necessary to lookup a symbol's value in `smt2_incremental_decision_proceduret::define_dependent_functions`?!

thomasspriggs replied:
Because the API for `decision_proceduret` is not sufficiently well documented or explicit with regards to how symbol expressions should be treated? I have (incorrectly) assumed that symbol expressions referring to symbols which have not previously been seen, means that these symbols are implicitly defined such that their value is equal to the value in the symbol table. This seemed to work well up to this point.

I can raise a PR which changes this such that symbol expressions are implicitly defined to have a nondeterministic value instead, if that is how a decision procedure should be implemented.

thomasspriggs commented:
The updating of `object_bits` in this PR looks to be incomplete. `config.bv_encoding.object_bits` is referenced in `simplify_exprt::simplify_pointer_offset`. The `simplify` function is called from various points before the analysis. Therefore the value of object bits could be embedded in various simplifications before the number of bits for the analysis is finalised.

tautschnig committed:
Do not use back-end pointer encoding details in the simplifier

The object:offset encoding is an implementation detail that should be
private to the back-end. Expression simplification must not rely on such
details. Partly reverts 4d35274bd0a.

tautschnig commented:
> The updating of `object_bits` in this PR looks to be incomplete. `config.bv_encoding.object_bits` is referenced in `simplify_exprt::simplify_pointer_offset`. The `simplify` function is called from various points before the analysis. Therefore the value of object bits could be embedded in various simplifications before the number of bits for the analysis is finalised.

Oh, good call-out. This seems just wrong. I have added a commit removing this piece of code.

thomasspriggs commented:
@tautschnig The regression tests for this PR should pass, once it is based on top of - https://github.com/diffblue/cbmc/pull/8104

tautschnig commented:
> @tautschnig The regression tests for this PR should pass, once it is based on top of - #8104

Rebased, thank you very much for the very quick fix!!

tautschnig commented:
CSmith test failure is down to a fix in our CSmith script being required since #8093 has been merged.

tautschnig commented:
> CSmith test failure is down to a fix in our CSmith script being required since #8093 has been merged.

Fixed in #8105.

peterschrammel commented on doc/man/goto-cc.1:
```diff
@@ -96,9 +96,10 @@ files.
-Configure the number of bits used for object numbering in CBMC's pointer encoding.
+.SH BACKWARD COMPATIBILITY
+.B goto\-cc
+will warn and ignore the use of \fB\-\-object\-bits\fR, which previous versions
```
❓ Where is this warning produced in this PR? Or does that already exist?

tautschnig replied:
goto-cc warns about all options that it doesn't explicitly know about, so by virtue of removing it from `gcc_cmdline.cpp` we get this.

tautschnig committed:
Mark pointer-in-range-requires test as expensive

This test appears to take several hours when running on Windows in
GitHub's CI.
